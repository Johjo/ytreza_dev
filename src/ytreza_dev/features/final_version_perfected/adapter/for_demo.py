from expression import Nothing
from pyqure import Key

from ytreza_dev.features.final_version_perfected.port.external_todolist import ExternalTodolistPort
from ytreza_dev.features.final_version_perfected.port.task_information_repository import TaskInformationRepositoryPort, \
    TaskInformation
from ytreza_dev.features.final_version_perfected.port.task_reader import TaskFvpReaderPort
from ytreza_dev.features.final_version_perfected.port.task_repository import FvpRepositoryPort
from ytreza_dev.features.final_version_perfected.types import TaskBase, ExternalTask


class TaskInMemory:
    def __init__(self) -> None:
        self._tasks: list[TaskBase] = []

    def all_tasks(self) -> list[TaskBase]:
        return self._tasks

    def save(self, tasks: list[TaskBase]) -> None:
        self._tasks = tasks


class TaskFvpReaderForDemo(TaskFvpReaderPort):
    def __init__(self, memory: TaskInMemory):
        self._memory = memory

    def all_active_tasks(self) -> list[TaskBase]:
        return self._memory.all_tasks()


class FvpRepositoryForDemo(FvpRepositoryPort):
    def all_tasks(self) -> list[TaskBase]:
        return self._memory.all_tasks()

    def __init__(self, memory: TaskInMemory):
        self._memory = memory

    def save(self, tasks: list[TaskBase]) -> None:
        self._memory.save(tasks)


TASK_IN_MEMORY_KEY = Key("task_in_memory", TaskInMemory)


class ExternalTodolistForDemo(ExternalTodolistPort):
    def close_task(self, url: str, task_id: str) -> None:
        pass


class TaskInformationRepositoryForDemo(TaskInformationRepositoryPort):
    def __init__(self):
        self._tasks: dict[str, TaskInformation] = {}

    def by_key(self, key: str) -> TaskInformation:
        return self._tasks[key]

    def save(self, task: list[TaskInformation]) -> None:
        for task in task:
            self._tasks[task.key] = task
