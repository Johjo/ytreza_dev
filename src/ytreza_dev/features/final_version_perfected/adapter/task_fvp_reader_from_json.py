import json
from pathlib import Path

from ytreza_dev.features.final_version_perfected.port.task_reader import TaskFvpReaderPort
from ytreza_dev.features.final_version_perfected.types import TaskBase, TaskNew, TaskNext, TaskLater, TaskNever, Project


class TaskFvpReaderFromJson(TaskFvpReaderPort):
    def __init__(self, json_path: Path) -> None:
        self._json_path = json_path

    def all_active_tasks(self) -> list[TaskBase]:
        tasks = json.loads(self._json_path.read_text(encoding="utf-8"))
        return [self._to_task(task) for task in tasks]

    def _to_task(self, task: dict[str, str]) -> TaskBase:
        match task["status"]:
            case "new":
                return TaskNew(title=task["title"], url=task["url"], id=task["id"])
            case "next":
                return TaskNext(title=task["title"], url=task["url"], id=task["id"])
            case "later":
                return TaskLater(title=task["title"], url=task["url"], id=task["id"])
            case "never":
                return TaskNever(title=task["title"], url=task["url"], id=task["id"])
        raise ValueError(f"Unknown status: {task['status']}")

    @staticmethod
    def _to_project(task) -> Project:
        return Project(key=task["project"]["key"], name=task["project"]["name"])

