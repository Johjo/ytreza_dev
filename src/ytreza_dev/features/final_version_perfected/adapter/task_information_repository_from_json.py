import json
from dataclasses import asdict
from pathlib import Path
from typing import Any

from expression import Nothing

from ytreza_dev.features.final_version_perfected.port.task_information_repository import TaskInformationRepositoryPort, \
    TaskInformation
from ytreza_dev.features.final_version_perfected.types import Project


class TaskInformationRepositoryFromJson(TaskInformationRepositoryPort):
    def __init__(self, json_path: Path) -> None:
        self._json_path = json_path


    def save(self, tasks: list[TaskInformation]) -> None:
        tasks_dict = [self._to_dict(task) for task in tasks]
        self._json_path.write_text(json.dumps(tasks_dict, indent=4))

    @staticmethod
    def _to_dict(task: TaskInformation) -> dict[str, Any]:
        d = asdict(task)
        return d

    def by_key(self, key: str) -> TaskInformation:
        tasks = json.loads(self._json_path.read_text(encoding="utf-8"))
        return [self._to_task_information(task) for task in tasks if task["key"] == key][0]

    @staticmethod
    def _to_task_information(task) -> TaskInformation:
        return TaskInformation(
            key=task["key"],
            title=task["title"],
            project=Project(key=task["project"]["key"], name=task["project"]["name"]),
            url=task["url"],
            due_date=Nothing
        )
