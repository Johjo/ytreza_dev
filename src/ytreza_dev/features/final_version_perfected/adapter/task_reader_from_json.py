import json
from pathlib import Path

from ytreza_dev.features.final_version_perfected.port.task_reader import TaskReader
from ytreza_dev.features.final_version_perfected.types import TaskBase, TaskNew, TaskNext, TaskLater, TaskNever


class TaskReaderFromJson(TaskReader):
    def __init__(self, json_path: Path) -> None:
        self._json_path = json_path

    def all_active_tasks(self) -> list[TaskBase]:
        tasks = json.loads(self._json_path.read_text(encoding="utf-8"))
        return [self._to_task(task) for task in tasks]

    @staticmethod
    def _to_task(task: dict[str, str]) -> TaskBase:
        match task["status"]:
            case "new":
                return TaskNew(title=task["title"], url=task["url"])
            case "next":
                return TaskNext(title=task["title"], url=task["url"])
            case "later":
                return TaskLater(title=task["title"], url=task["url"])
            case "never":
                return TaskNever(title=task["title"], url=task["url"])
        raise ValueError(f"Unknown status: {task['status']}")

