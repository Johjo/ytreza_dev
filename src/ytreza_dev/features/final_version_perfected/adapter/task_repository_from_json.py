import json
from dataclasses import asdict
from pathlib import Path
from typing import Any

from ytreza_dev.features.final_version_perfected.port.task_repository import FvpRepositoryPort
from ytreza_dev.features.final_version_perfected.types import TaskBase, TaskNew, TaskNext, TaskLater, TaskNever


class FvpRepositoryFromJson(FvpRepositoryPort):
    def all_tasks(self) -> list[TaskBase]:

        tasks = json.loads(self._file_path.read_text())
        return [self.to_task_base(task) for task in tasks]

    @staticmethod
    def to_task_base(task: dict[str, str]) -> TaskBase:
        match task["status"]:
            case "new":
                return TaskNew(id=task["id"])
            case "next":
                return TaskNext(id=task["id"])
            case "later":
                return TaskLater(id=task["id"])
            case "never":
                return TaskNever(id=task["id"])
        raise ValueError(f"Unknown status: {task['status']}")

    def __init__(self, file_path: Path) -> None:
        self._file_path = file_path

    def save(self, tasks: list[TaskBase]) -> None:
        tasks_dict = [self._to_dict(task) for task in tasks]
        self._file_path.write_text(json.dumps(tasks_dict, indent=4))

    @staticmethod
    def _to_dict(task: TaskBase) -> dict[str, Any]:
        d = asdict(task)
        match task:
            case TaskNext():
                d.update({"status": "next"})
            case TaskNew():
                d.update({"status": "new"})
            case TaskLater():
                d.update({"status": "later"})
            case TaskNever():
                d.update({"status": "never"})

        return d