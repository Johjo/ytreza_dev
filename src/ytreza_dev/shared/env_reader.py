import os
from abc import ABC


class EnvReaderPort(ABC):
    def read(self, key) -> str:
        value = os.getenv(key)
        if value is None:
            raise KeyError(f"env key {key} is not defined")
        return value


class EnvReaderFromEnv(EnvReaderPort):
    pass
