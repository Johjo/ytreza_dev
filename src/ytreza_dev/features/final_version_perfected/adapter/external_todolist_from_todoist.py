from ytreza_dev.features.final_version_perfected.port.external_todolist import ExternalTodolistPort
from ytreza_dev.shared.env_reader import EnvReaderPort
from ytreza_dev.shared.todoist.todoist_api import TodoistAPI


class ExternalTodolistFromTodoist(ExternalTodolistPort):
    def __init__(self, env_reader: EnvReaderPort) -> None:
        self._todoist = TodoistAPI(env_reader.read("TODOIST_API_TOKEN"))

    def close_task(self, url: str, task_id: str) -> None:
        print(f"close task {task_id} on {url}")
        self._todoist.close_task(task_id=task_id)
