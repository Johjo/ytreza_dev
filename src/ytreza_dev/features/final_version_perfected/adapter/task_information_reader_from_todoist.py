from ytreza_dev.features.final_version_perfected.port.task_information_reader import TaskInformationReaderPort, \
    TaskInformation
from ytreza_dev.features.final_version_perfected.types import Project
from ytreza_dev.shared.env_reader import EnvReaderFromEnv
from ytreza_dev.shared.todoist.todoist_api import TodoistAPI, TodoistTask


class TaskInformationReaderFromTodoist(TaskInformationReaderPort):
    def __init__(self, env_reader: EnvReaderFromEnv) -> None:
        self._todoist = TodoistAPI(env_reader.read("TODOIST_API_TOKEN"))

    def by_key(self, key: str) -> TaskInformation:
        task : TodoistTask = self._todoist.task_by_id(task_id=key)
        return TaskInformation(key=task.id, title=task.name, project=Project(key=task.project.id, name=task.project.name), due_date=None)
