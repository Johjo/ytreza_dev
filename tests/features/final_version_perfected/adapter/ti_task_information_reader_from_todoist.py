from approvaltests import verify

from ytreza_dev.features.final_version_perfected.adapter.task_information_reader_from_todoist import \
    TaskInformationReaderFromTodoist
from ytreza_dev.shared.env_reader import EnvReaderFromEnv


def test_read_information() -> None:
    env_reader = EnvReaderFromEnv(env_path=".env.test")

    sut = TaskInformationReaderFromTodoist(env_reader=env_reader)
    information = sut.by_key("9497217837")

    verify(information, encoding="utf-8")
