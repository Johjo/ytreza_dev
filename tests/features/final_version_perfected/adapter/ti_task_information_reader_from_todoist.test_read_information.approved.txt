TaskInformation(key='9388991121', title='Téléchargez des [applications et plugins gratuits supplémentaires](https://todoist.com/fr/downloads?utm_source=todoist&utm_medium=in_app&utm_campaign=onboarding_project&utm_content=inbox) pour 💻,⌚️,🖥,📱 et 📧', project=Project(key='2357501003', name='Inbox'), due_date=None)
